File: app/src/main/java/com/example/messengerandroid/fragment/SignInFragment.java
package com.example.messengerandroid.fragment;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.util.Patterns;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.example.messengerandroid.MainActivity;
import com.example.messengerandroid.R;
import com.example.messengerandroid.databinding.ActivitySignInBinding;
import com.example.messengerandroid.model.User;

import java.util.HashMap;

public class SignInFragment extends Fragment {

    private static final String TAG = "SignInFragment";

    private ActivitySignInBinding binding;

    private HashMap<String, String> users = new HashMap<>();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = ActivitySignInBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Log.d(TAG, "onCreate");
        if (getArguments() != null) {
            User user = SignInFragmentArgs.fromBundle(getArguments()).getUser();
            if (user != null) {
                users.put(user.getEmail(), user.getPassword());
                binding.etEmail.setText(user.getEmail());
                Toast.makeText(getActivity(), "Регистрация успешна!\nИмя: " + user.getUsername()
                        + "\nEmail: " + user.getEmail(), Toast.LENGTH_SHORT).show();
            }
        }

        binding.btnSignIn.setOnClickListener(v -> {
            String emailInput = binding.etEmail.getText().toString().trim();
            String passwordInput = binding.etPassword.getText().toString().trim();

            if (validateInput(emailInput, passwordInput)) {
                if (passwordInput.equals(users.get(emailInput))) {
                    ((MainActivity) requireActivity()).navigateToHome();
                } else {
                    Toast.makeText(getActivity(), "Неверный email или пароль", Toast.LENGTH_SHORT).show();
                }
            }
        });

        binding.btnRegister.setOnClickListener(v -> ((MainActivity) requireActivity()).navigateToSignUp());
    }

    private boolean validateInput(String email, String password) {
        if (TextUtils.isEmpty(email)) {
            binding.etEmail.setError("Введите email");
            return false;
        }

        if (!Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            binding.etEmail.setError("Некорректный формат email");
            return false;
        }

        if (TextUtils.isEmpty(password)) {
            binding.etPassword.setError("Введите пароль");
            return false;
        }

        if (password.length() < 6) {
            binding.etPassword.setError("Пароль должен содержать минимум 6 символов");
            return false;
        }

        return true;
    }
}

File: app/src/main/java/com/example/messengerandroid/fragment/SignInFragment.kt
package com.example.messengerandroid.fragment

import android.os.Bundle
import android.util.Log
import android.util.Patterns
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.example.messengerandroid.MainActivity
import com.example.messengerandroid.R
import com.example.messengerandroid.databinding.ActivitySignInBinding

class SignInFragment : Fragment() {

    private var binding: ActivitySignInBinding? = null

    private val users = hashMapOf<String, String>()

    companion object {
        private const val TAG = "SignInFragment"
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = ActivitySignInBinding.inflate(inflater, container, false)
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        Log.d(TAG, "onCreate")

        arguments?.let { bundle ->
            val user = SignInFragmentArgs.fromBundle(bundle).user
            user?.let {
                users[it.email] = it.password
                binding?.etEmail?.setText(it.email)
                Toast.makeText(
                    activity,
                    getString(R.string.registration_success, it.username, it.email),
                    Toast.LENGTH_SHORT
                ).show()
            }
        }

        binding?.apply {
            btnSignIn.setOnClickListener {
                val emailInput = etEmail.text.toString().trim()
                val passwordInput = etPassword.text.toString().trim()

                if (validateInput(emailInput, passwordInput)) {
                    if (users[emailInput] == passwordInput) {
                        (requireActivity() as MainActivity).navigateToCharacters()
                    } else {
                        Toast.makeText(
                            activity,
                            getString(R.string.invalid_email_password),
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                }
            }

            btnRegister.setOnClickListener {
                (requireActivity() as MainActivity).navigateToSignUp()
            }
        }
    }

    private fun validateInput(email: String, password: String): Boolean {
        var isValid = true

        binding?.apply {
            if (email.isEmpty()) {
                etEmail.error = getString(R.string.enter_email)
                isValid = false
            } else if (!Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
                etEmail.error = getString(R.string.invalid_email_format)
                isValid = false
            }

            if (password.isEmpty()) {
                etPassword.error = getString(R.string.enter_password)
                isValid = false
            } else if (password.length < 6) {
                etPassword.error = getString(R.string.password_too_short)
                isValid = false
            }
        }

        return isValid
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }
}
